### PR TITLE
Restructure climb into climb and retract, add manual override

### PR DESCRIPTION
**Pull request by SaltyJoke**
Not quite related: either Drivetrain method isStalled() is badly named or the logic in Climb execute() is wrong. Can't tell which with what little understanding I have. Currently the up-down phase ends when isStalled() is false.

### CODE REVIEW

File: Robot2019/src/main/java/frc/robot/commands/Climb.java [hunk 1]
```diff
@@ -33,28 +38,30 @@ protected void initialize() {
...
   @Override
   protected void execute() {
-    dt.drive(-0.5, -0.5);
-    if (up) {
-      climber.runClimber(0.5);
-      if (!climber.needToClimb())
-        up = false;
+    if (state == State.CLIMB) {
+      climb();
+      if (!dt.isStalled() || Math.abs(dtJoy.getY()) > 0.1) { // 0.1 = Placeholder/approximation
```
Comment (kevinzwang) [inline Robot2019/src/main/java/frc/robot/commands/Climb.java]: what is the joystick supposed to do?

File: Robot2019/src/main/java/frc/robot/commands/Climb.java [hunk 2]
```diff
@@ -33,28 +38,30 @@ protected void initialize() {
...
+        climber.resetEncoder();
+      }
     } else {
-      climber.runClimber(-0.5);
-      if (!climber.canDrop())
-        up = true;
+      retract();
+      if (climber.getEncDistance() > 23) { // 23 = Placeholder/approximation 
+        state = State.FINISHED;
+      }
```
Comment (kevinzwang) [inline Robot2019/src/main/java/frc/robot/commands/Climb.java]: This code only goes up and then down once. Is that what we want to do?

File: Robot2019/src/main/java/frc/robot/commands/Climb.java [hunk 3]
```diff
@@ -33,28 +38,30 @@ protected void initialize() {
...
     }
   }
 
   // Make this return true when this Command no longer needs to run execute()
   @Override
   protected boolean isFinished() {
-    return !dt.isStalled();
+    return state == state.FINISHED;
```
Comment (kevinzwang) [inline Robot2019/src/main/java/frc/robot/commands/Climb.java]: wrong State

File: Robot2019/src/main/java/frc/robot/commands/Climb.java [hunk 4]
```diff
@@ -63,4 +70,27 @@ protected void end() {
...
+    if (up) {
+      climber.runClimber(0.5);
+      if (!climber.needToClimb())
+        up = false;
+    } else {
+      climber.runClimber(-0.5);
+      if (!climber.canDrop())
+        up = true;
+    }
+  }
```
Comment (kevinzwang) [inline Robot2019/src/main/java/frc/robot/commands/Climb.java]: this code does not work with the current execute() method

File: Robot2019/src/main/java/frc/robot/commands/Climb.java [hunk 5]
```diff
@@ -63,4 +70,27 @@ protected void end() {
...
+
+  private void retract() {
+    climber.retractClimber();
+  }
```
Comment (kevinzwang) [inline Robot2019/src/main/java/frc/robot/commands/Climb.java]: either keep both the climb() and retract() logic all in the Climber subsystem class or all in here, preferably in the subsystem

File: Robot2019/src/main/java/frc/robot/commands/Climb.java (hunk 2 above)
Comment (SaltyJoke) [reply Robot2019/src/main/java/frc/robot/commands/Climb.java]: That part only occurs after the stage moves to RETRACT.

File: Robot2019/src/main/java/frc/robot/commands/Climb.java (hunk 4 above)
Comment (SaltyJoke) [reply Robot2019/src/main/java/frc/robot/commands/Climb.java]: I don't know how to fix that

**Comment by alexander-mcdowell** [comment]
Math.atan2 returns theta between y and x in RADIANS, not degrees. Choose to either use degrees or radians for maxTilt (I would suggest radians).

**Comment by SaltyJoke** [comment]
> Math.atan2 returns theta between y and x in RADIANS, not degrees. Choose to either use degrees or radians for maxTilt (I would suggest radians).

Darn, I'm pretty sure I'd taken care of that initially but it must have been lost when David rewrote it.

**Comment by alexander-mcdowell** [comment]
My bad sorry

File: Robot2019/src/main/java/frc/robot/commands/Climb.java [hunk 6]
```diff
@@ -10,19 +10,24 @@
...
 
-  public Climb(Climber climber, Drivetrain dt) {
+  public Climb(Climber climber, Drivetrain dt, Joystick joy) {
     // Use requires() here to declare subsystem dependencies
     this.climber = climber;
     this.dt = dt;
+    dtJoy = joy;
     requires(climber);
     requires(dt);
     up = true;
```
Comment (alexander-mcdowell) [inline Robot2019/src/main/java/frc/robot/commands/Climb.java]: It does not seem like the variable "up" is ever being used in Climb.java. Is this variable supposed to be replaced with the state machine?

Comment (SaltyJoke) [reply Robot2019/src/main/java/frc/robot/commands/Climb.java]: up is used in climb()

File: Robot2019/src/main/java/frc/robot/commands/Climb.java (hunk 1 above)
Comment (SaltyJoke) [reply Robot2019/src/main/java/frc/robot/commands/Climb.java]: Checks if driver is trying to drive. Not sure what to fix.

**Comment by SaltyJoke** [comment]
I may have misunderstood or failed to understand some of Kevin's requests.

File: Robot2019/src/main/java/frc/robot/commands/Climb.java (hunk 6 above)
Comment (alexander-mcdowell) [reply Robot2019/src/main/java/frc/robot/commands/Climb.java]: My bad, I did not see that.

File: Robot2019/src/main/java/frc/robot/commands/Climb.java (hunk 1 above)
Comment (kevinzwang) [reply Robot2019/src/main/java/frc/robot/commands/Climb.java]: I'd suggest a button override instead

File: Robot2019/src/main/java/frc/robot/commands/Climb.java [hunk 7]
```diff
@@ -63,4 +83,19 @@ protected void end() {
   protected void interrupted() {
     end();
   }
+
+  private void checkForRetracting() {
+    double joyMovingThreshold = 0.1; // Test for actual number
+      if (!dt.isStalled() || Math.abs(dtJoy.getY()) > joyMovingThreshold) {
```
Comment (brettle) [inline Robot2019/src/main/java/frc/robot/commands/Climb.java]: The isStalled logic looks backward here.

**Comment by doawelul** [comment]
This also fixes #28

**Pull request merged**